mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesService.java

```diff
@@ -0,0 +1,437 @@
+            this.updatedMessageIdEntries = updatedMessageIdEntries;
+            this.removedMessageIdEntries = removedMessageIdEntries;
+            this.fixedInconsistencies = new ConcurrentLinkedDeque<>(fixedInconsistencies);
+            this.errors = errors;
+        }
+
+        void incrementProcessedImapUidEntries() {
+            processedImapUidEntries.incrementAndGet();
+        }
+
+        void incrementMessageIdEntries() {
+            processedMessageIdEntries.incrementAndGet();
+        }
+
+        void incrementAddedMessageIdEntries() {
+            addedMessageIdEntries.incrementAndGet();
+        }
+
+        void incrementUpdatedMessageIdEntries() {
+            updatedMessageIdEntries.incrementAndGet();
+        }
+
+        void incrementRemovedMessageIdEntries() {
+            removedMessageIdEntries.incrementAndGet();
+        }
+
+        void addFixedInconsistency(ComposedMessageId messageId) {
+            fixedInconsistencies.add(messageId);
+        }
+
+        void incrementErrors() {
+            errors.incrementAndGet();
+        }
+
+        Snapshot snapshot() {
+            return new Snapshot(
+                processedImapUidEntries.get(),
+                processedMessageIdEntries.get(),
+                addedMessageIdEntries.get(),
+                updatedMessageIdEntries.get(),
+                removedMessageIdEntries.get(),
+                ImmutableList.copyOf(fixedInconsistencies),
+                errors.get());
+        }
+    }
+
+    public static final Logger LOGGER = LoggerFactory.getLogger(SolveMessageInconsistenciesService.class);
+
+    private final CassandraMessageIdToImapUidDAO messageIdToImapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+
+    @Inject
+    SolveMessageInconsistenciesService(CassandraMessageIdToImapUidDAO messageIdToImapUidDAO, CassandraMessageIdDAO messageIdDAO) {
+        this.messageIdToImapUidDAO = messageIdToImapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+    }
+
+    Mono<Task.Result> fixMessageInconsistencies(Context context) {
+        return Flux.concat(
+            fixInconsistenciesInMessageId(context),
+            fixInconsistenciesInImapUid(context))
+            .reduce(Task.Result.COMPLETED, Task::combine);
+    }
+
+    private Flux<Task.Result> fixInconsistenciesInImapUid(Context context) {
+        return messageIdToImapUidDAO.retrieveAllMessages()
+            .doOnNext(any -> context.incrementProcessedImapUidEntries())
+            .concatMap(this::detectInconsistencyInImapUid)
+            .concatMap(inconsistency -> inconsistency.fix(context, messageIdToImapUidDAO, messageIdDAO));
+    }
+
+    private Mono<Inconsistency> detectInconsistencyInImapUid(ComposedMessageIdWithMetaData message) {
+        return messageIdToImapUidDAO.retrieve((CassandraMessageId) message.getComposedMessageId().getMessageId(), Optional.of((CassandraId) message.getComposedMessageId().getMailboxId()))
```

Comment:
other methods are quite easy to read, you did great, this one is a bit too technical however, it would deserve some code extractions to make it more readable